ql/src/java/org/apache/hadoop/hive/ql/schq/ScheduledQueryExecutionService.java

```diff
@@ -0,0 +1,159 @@
+    ExecutorService executor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder().setDaemon(true).setNameFormat("Scheduled Query Thread %d").build());
+    ScheduledQueryExecutionContext ctx = new ScheduledQueryExecutionContext(executor, conf, qService);
+    return new ScheduledQueryExecutionService(ctx);
+  }
+
+  public ScheduledQueryExecutionService(ScheduledQueryExecutionContext ctx) {
+    context = ctx;
+    ctx.executor.submit(worker = new ScheduledQueryExecutor());
+    ctx.executor.submit(new ProgressReporter());
+  }
+
+  class ScheduledQueryExecutor implements Runnable {
+
+    private ScheduledQueryProgressInfo info;
+
+    @Override
+    public void run() {
+      while (true) {
+        ScheduledQueryPollResponse q = context.schedulerService.scheduledQueryPoll();
+        if (q.isSetExecutionId()) {
+          try{
+            processQuery(q);
+          } catch (Throwable t) {
+            LOG.error("Unexpected exception during scheduled query processing", t);
+          }
+        } else {
+          try {
+            Thread.sleep(context.getIdleSleepTime());
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOG.info("interrupted");
+            break;
+          }
+        }
+      }
+    }
+
+    QueryState lastState;
+    public synchronized void reportQueryProgress() {
+      if (info != null) {
+        if(lastState == QueryState.FINISHED || lastState == QueryState.ERRORED) {
+          return;
+        }
+        LOG.info("Reporting query progress of {} as {} err:{}", info.getScheduledExecutionId(), info.getState(),
+            info.getErrorMessage());
+        context.schedulerService.scheduledQueryProgress(info);
+        lastState=info.getState();
+      }
+    }
+
+    private void processQuery(ScheduledQueryPollResponse q) {
+      SessionState state = null;
+      try {
+        state = SessionState.start(context.conf);
+        info = new ScheduledQueryProgressInfo();
+        info.setScheduledExecutionId(q.getExecutionId());
+        info.setState(QueryState.EXECUTING);
+        reportQueryProgress();
+        try (
+          IDriver driver = DriverFactory.newDriver(DriverFactory.getNewQueryState(context.conf), q.getUser(), null)) {
+          info.setExecutorQueryId(driver.getQueryState().getQueryId());
+          CommandProcessorResponse resp;
+          resp = driver.run(q.getQuery());
+          if (resp.getResponseCode() != 0) {
+            throw resp;
+          }
+          info.setState(QueryState.FINISHED);
+        }
+      } catch (Throwable t) {
+        info.setErrorMessage(getErrorStringForException(t));
+        info.setState(QueryState.ERRORED);
+      } finally {
+        if (state != null) {
+          try {
+            state.close();
+          } catch (Throwable e) {
+          }
+        }
+
+        synchronized (this) {
```

Comment:
Wouldn't it be easier to set it from `reportQueryProgress` since it is already synchronized (you could pass a `clearInfo` boolean to the method)? Up to you, I just find it easier in general to reason about the code minimizing the number of synchronized code blocks :)